src/KeyManager.ts

```diff
@@ -340,68 +353,56 @@ export class KeyManager {
 
     const text = await challengeRes.text();
 
-    let transaction;
-    let network_passphrase;
-    let error;
+    let json;
 
     try {
-      const json = JSON.parse(text);
-
-      transaction = json.transaction;
-      network_passphrase = json.network_passphrase;
-      error = json.error;
+      json = JSON.parse(text);
     } catch (e) {
       throw new Error(`Request for challenge returned invalid JSON: ${text}`);
     }
 
-    if (error) {
-      throw new Error(error);
+    if (json.error) {
+      throw new Error(json.error);
     }
 
     // Throw error when network_passphrase is returned, and doesn't match
-    if (network_passphrase !== undefined && keyNetwork !== network_passphrase) {
+    if (
+      json.network_passphrase !== undefined &&
+      keyNetwork !== json.network_passphrase
+    ) {
       throw new Error(
         `
-        Network mismatch: the transfer server expects "${network_passphrase}",
-        but you're using "${keyNetwork}"
-        `,
+            Network mismatch: the transfer server expects "${
+              json.network_passphrase
+            }",
+            but you're using "${keyNetwork}"
+            `,
       );
     }
 
-    const keyHandler = this.keyHandlerMap[key.type];
+    const firstTransaction = Utils.readChallengeTx(
+      json.transaction,
+      authServerKey,
+      keyNetwork,
+      authServerHomeDomains,
+      new URL(authServer).hostname,
+    ).tx;
 
-    const firstTransaction = new Transaction(transaction, keyNetwork);
+    const keyHandler = this.keyHandlerMap[key.type];
 
-    if (firstTransaction.sequence !== "0") {
+    if (
+      !firstTransaction.signatures.some((signature) =>
+        signature
+          .hint()
+          .equals(
+            StellarSdk.Keypair.fromPublicKey(authServerKey).signatureHint(),
+          ),
+      )
+    ) {
       throw new Error(
-        `Invalid transaction: Expected a sequence number 0, but got ${
-          firstTransaction.sequence
-        }`,
-      );
-    }
-
-    if (authServerKey) {
-      if (firstTransaction.source !== authServerKey) {
-        throw new Error(
-          `Signing key doesn't match: Expected ${authServerKey} but got
-          ${firstTransaction.source}`,
-        );
-      }
-
-      if (
-        !firstTransaction.signatures.some((signature) =>
-          signature
-            .hint()
-            .equals(
-              StellarSdk.Keypair.fromPublicKey(authServerKey).signatureHint(),
-            ),
-        )
-      ) {
-        throw new Error(
-          `Signing key doesn't match: Expected ${authServerKey} but got
+        `Signing key doesn't match: Expected ${authServerKey} but got
```

Comment:
@morleyzhi @fnando I don't think you need this error message because in v8.2.3+ the `readChallengeTx` will throw an error if not signed by the server.